### ISSUE TITLE
add Iterator to compliment ReverseIterator

### ISSUE THREAD

**Issue by ic3man5** [ziglang/zig#22332]
### Zig Version

0.13.0

### Steps to Reproduce and Observed Behavior

Currently the standard library has `reverseIterator` but doesn't have something similar for a "Normal" Iterator.

Below is code to show how this would work, if accepted I could make a PR for this.

```zig

const std = @import("std");
const testing = std.testing;

pub fn main() !void {
    const alphabet = "abcdefghijklmnopqrstuvwxyz";
    // Normal Iterator
    var it = iterator(alphabet);
    displayIterator(@TypeOf(it), &it);
    // Reverse Iterator
    var r_it = std.mem.reverseIterator(alphabet);
    displayIterator(@TypeOf(r_it), &r_it);
}

fn displayIterator(comptime T: type, iter: *T) void {
    while (iter.next()) |i| {
        std.debug.print("{c} ", .{i});
    }
    std.debug.print("\n", .{});
}

fn Iterator(comptime T: type) type {
    const Pointer = blk: {
        switch (@typeInfo(T)) {
            .Pointer => |ptr_info| switch (ptr_info.size) {
                .One => switch (@typeInfo(ptr_info.child)) {
                    .Array => |array_info| {
                        var new_ptr_info = ptr_info;
                        new_ptr_info.size = .Many;
                        new_ptr_info.child = array_info.child;
                        new_ptr_info.sentinel = array_info.sentinel;
                        break :blk @Type(.{ .Pointer = new_ptr_info });
                    },
                    else => {},
                },
                .Slice => {
                    var new_ptr_info = ptr_info;
                    new_ptr_info.size = .Many;
                    break :blk @Type(.{ .Pointer = new_ptr_info });
                },
                else => {},
            },
            else => {},
        }
        @compileError("expected slice or pointer to array, found '" ++ @typeName(T) ++ "'");
    };
    const Element = std.meta.Elem(Pointer);
    const ElementPointer = @Type(.{ .Pointer = ptr: {
        var ptr = @typeInfo(Pointer).Pointer;
        ptr.size = .One;
        ptr.child = Element;
        ptr.sentinel = null;
        break :ptr ptr;
    } });
    return struct {
        ptr: Pointer,
        index: usize,
        len: usize,
        pub fn next(self: *@This()) ?Element {
            if (self.index == self.len) return null;
            self.index += 1;
            return self.ptr[self.index - 1];
        }
        pub fn nextPtr(self: *@This()) ?ElementPointer {
            if (self.index == self.len) return null;
            self.index += 1;
            return &self.ptr[self.index - 1];
        }
        pub fn reset(self: *@This()) void {
            self.index = 0;
        }
    };
}

/// Iterates over a slice.
pub fn iterator(slice: anytype) Iterator(@TypeOf(slice)) {
    return .{ .ptr = slice.ptr, .index = 0, .len = slice.len };
}
```

### Expected Behavior

Standard Library to have an Iterator.

**Comment by ic3man5** [comment]
Output of code above:

```bash
$ zig build run
a b c d e f g h i j k l m n o p q r s t u v w x y z 
z y x w v u t s r q p o n m l k j i h g f e d c b a 
```

**Comment by tauoverpi** [comment]
Might seem like a silly question but why not just a for loop?

**Comment by ic3man5** [comment]
> Might seem like a silly question but why not just a for loop?

I found myself wanting to do a normal iteration of a slice and a reverse iteration of a slice. The for loop would work but if you want a common function to iterate both ways you can't really use the reverse iterator. This is why I included `displayIterator` in the code above.

I'm not sure what the intended goal of the standard library is but if its to include iterators then this addition makes sense to me since we already have the reverseIterator and others like split function in `std.mem`. I'm open to suggestions.

**Comment by ic3man5** [comment]
Also coming from languages like C++ and Rust have iterators so it seems like a good feature to have IMO. Go recently added iterator support in the standard library also [here](https://github.com/golang/go/issues/61897). I would consider Go and Zig in the same classification of "simple" languages.

**Comment by andrewrk** [comment]
> a common function to iterate both ways

This is not compelling.

**Comment by ic3man5** [comment]
> > a common function to iterate both ways
> 
> This is not compelling.

Should I open a PR to remove the `reverseIterator` then? It doesn't make sense to me to have one but not the other?

**Comment by mlugg** [comment]
No, because `reverseIterator` does something which there isn't a language-level feature for. If you want to iterate a sequence in order, use a `for` loop; this is a solved problem.

**Comment by andrewrk** [comment]
If you're looking for something to do, I recommend browsing the "contributor friendly" label.

**Comment by ic3man5** [comment]
> No, because `reverseIterator` does something which there isn't a language-level feature for. If you want to iterate an array in order, use a `for` loop; this is a solved problem.

Is there a better way to do this?

Proposed:
```zig
fn displayIterator(comptime T: type, iter: *T) void {
    while (iter.next()) |i| {
        std.debug.print("{c} ", .{i});
    }
    std.debug.print("\n", .{});
}```

Current:
```zig
fn displaySlice(comptime T: type, data: []const T, reverse: bool) void {
    if (reverse) {
        var iter = std.mem.reverseIterator(data);
        while (iter.next()) |i| {
            std.debug.print("{c} ", .{i});
        }
    } else {
        for (data) |b| {
            std.debug.print("{c} ", .{b});
        }
    }
    std.debug.print("\n", .{});
}```

**Comment by RetroDev256** [comment]
Seems like an overly abstracted function. I do understand that this example is derived to explain the need of a generic and similar way to iterate both ways, but it is not common enough to need one function to handle both methods of iteration, or even just reverse iteration in my opinion. 99% of slices will be iterated forward.

**Comment by mlugg** [comment]
Zig discourages unnecessarily generic code like that example. If you really need it, you can always include a forwards iterator in your own codebase, but it's quite unlikely that you *actually* need it.

**Comment by andrewrk** [comment]
Of those two examples, "Current" looks like better code to me. I would refactor the "Proposed" into "Current" if I saw it and was working on related code.

Better yet, you can further refactor "Current" into:

```zig
fn displaySliceForward(comptime T: type, data: []const T) void {
    for (data) |b| {
        std.debug.print("{c} ", .{b});
    }
    std.debug.print("\n", .{});
}

fn displaySliceReverse(comptime T: type, data: []const T) void {
    var iter = std.mem.reverseIterator(data);
    while (iter.next()) |i| {
        std.debug.print("{c} ", .{i});
    }
    std.debug.print("\n", .{});
}
```

I've removed all the abstraction and now look how your code is improved!

```diff
-    // Normal Iterator
-    var it = iterator(alphabet);
-    displayIterator(@TypeOf(it), &it);
-    // Reverse Iterator
-    var r_it = std.mem.reverseIterator(alphabet);
-    displayIterator(@TypeOf(r_it), &r_it);
+    // Normal Iterator
+    displaySliceForward(u8, alphabet);
+    // Reverse Iterator
+    displaySliceReverse(u8, alphabet);
```

[Recommended reading](https://caseymuratori.com/blog_0015)

**Comment by ic3man5** [comment]
> Of those two examples, "Current" looks like better code to me. I would refactor the "Proposed" into "Current" if I saw it and was working on related code.
> 
> Better yet, you can further refactor "Current" into:
> 
> ```zig
> fn displaySliceForward(comptime T: type, data: []const T) void {
>     for (data) |b| {
>         std.debug.print("{c} ", .{b});
>     }
>     std.debug.print("\n", .{});
> }
> 
> fn displaySliceReverse(comptime T: type, data: []const T) void {
>     var iter = std.mem.reverseIterator(data);
>     while (iter.next()) |i| {
>         std.debug.print("{c} ", .{i});
>     }
>     std.debug.print("\n", .{});
> }
> ```
> 

Wouldn't it be even better if you could compact this further by only having one function (which having an implementation of `Iterator` would achieve? Coming from other projects (typically written in C for example), a lot of quick and dirty hacks I've seen when the language doesn't allow simple things like this to happen the lazy programmer (or the programmer on time constraints) will just copy and paste (and modify one thing) and eventually we end up with an unmaintainable nightmare.

I see the above as a copy and paste; What if we need another function that goes Forward by 2, or back by 2 (for example)? In the future you'll potentially have more copy/paste. Now if we need to modify the behavior of one we now need to modify it everywhere leading to errors.

Keep in mind the only reason we need two functions above is because Zig doesn't have language constructs to handle forward and reverse in the same way. The only change above is changing the loop semantics from `for` to `while`.

I see both sides of the arguments and agree with both. My main concern is it puts too much of a "C" taste in my mouth where I'm "fighting" the language over actually getting stuff done.

**Comment by mlugg** [comment]
> What if we need another function that goes Forward by 2, or back by 2 (for example)?

The recommended reading provided by Andrew encourages you to move away from this thought process, and with good reason. Some programmers spend so much of their time asking "what if I need X?" that it gets completely in the way of them writing actual code!

If down the line you end up realising your project needs all manner of iteration behaviours for some reason, and it would be clearer to handle them generically, then by all means, go ahead and implement a generic iterator which can go in either direction and jump by an arbitrary amount. However, in reality, you almost certainly *don't* need this -- so why spend effort on it? If you just need to iterate forwards, or even just forwards and backwards, then trying to generalize the concept isn't helping anyone: it wastes your time as the author, and it wastes the time of anyone reading your code, who now has to grasp the meaning of some funky generic logic.

**Comment by ic3man5** [comment]
> > What if we need another function that goes Forward by 2, or back by 2 (for example)?
> 
> The recommended reading provided by Andrew encourages you to move away from this thought process, and with good reason. Some programmers spend so much of their time asking "what if I need X?" that it gets completely in the way of them writing actual code!
> 

I was referring to the second part of the article where you find yourself having to refactor because of a lot of copy and pasted code. In this instance and the reason I looked into iterator in the first place is because I was attempting to refactor copy and pasted code. I've been trapped inside prematurely designing code like that a bunch of times to the point where I'd like to think I don't do it anymore. The first part of the article example with inheritance screams just do composition instead, I overlooked that though because it wasn't trying to fix inheritance design.
